google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java

```diff
@@ -0,0 +1,568 @@
+    // should trigger the first batch to be sent.
+    bulkWriter.delete(firestoreMock.document("coll/doc4"));
+
+    List<BatchWriteRequest> requests = batchWriteCapture.getAllValues();
+    assertEquals(requestResponseMap.size(), requests.size());
+
+    verifyRequests(requests, requestResponseMap);
+    assertEquals(Timestamp.ofTimeSecondsAndNanos(0, 0), result1.get().getUpdateTime());
+    assertEquals(Timestamp.ofTimeSecondsAndNanos(1, 0), result2.get().getUpdateTime());
+    assertEquals(Timestamp.ofTimeSecondsAndNanos(2, 0), result3.get().getUpdateTime());
+  }
+
+  @Test
+  public void doesNotSendBatchesIfSameDocIsInFlight() throws Exception {
+    final RequestResponseMap requestResponseMap =
+        new RequestResponseMap() {
+          {
+            put(
+                batchWrite(
+                    set(LocalFirestoreHelper.SINGLE_FIELD_PROTO, "coll/doc1"),
+                    set(LocalFirestoreHelper.SINGLE_FIELD_PROTO, "coll/doc2")),
+                mergeResponses(successResponse(1), successResponse(2)));
+            put(
+                batchWrite(set(LocalFirestoreHelper.SINGLE_FIELD_PROTO, "coll/doc1")),
+                successResponse(3));
+          }
+        };
+    requestResponseMap.initializeStub(batchWriteCapture, firestoreMock, true);
+    bulkWriter.set(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+    bulkWriter.set(doc2, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+
+    // Schedule flush on separate thread to avoid blocking main thread while waiting for
+    // activeRequestComplete.
+    ScheduledFuture<ApiFuture<Void>> flush1 =
+        Executors.newSingleThreadScheduledExecutor()
+            .schedule(
+                new Callable<ApiFuture<Void>>() {
+                  public ApiFuture<Void> call() {
+                    return bulkWriter.flush();
+                  }
+                },
+                0,
+                TimeUnit.MILLISECONDS);
+
+    // Wait for flush() to perform logic and reach the stubbed response. This simulates a first
+    // batch that has been sent with its response still pending.
+    requestResponseMap.awaitLatch();
+
+    bulkWriter.set(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+    ApiFuture<Void> flush2 = bulkWriter.flush();
+
+    // Wait for flush() to receive its response and process the batch.
+    requestResponseMap.markComplete();
+    flush1.get().get();
+    flush2.get();
+    bulkWriter.close().get();
+
+    List<BatchWriteRequest> requests = batchWriteCapture.getAllValues();
+    assertEquals(requestResponseMap.size(), requests.size());
+
+    verifyRequests(requests, requestResponseMap);
+  }
+
+  @Test
+  public void doesNotSendBatchesIfDoingSoExceedsRateLimit() throws Exception {
+    // The test is considered a success if BulkWriter tries to send the second batch again after a
+    // timeout.
+
+    List<Integer> arrayRange = new ArrayList<>();
+    for (int i = 0; i < 500; ++i) {
```

Comment:
Can we "compute" this value inline?

Reply:
I don't see why we need it. There is only one place where we are looping over `arrayRange` and we could just use `for (int i = 0; i < 500; ++i)` instead.